scipy/sparse/linalg/_isolve/tests/test_iterative.py

```diff
@@ -325,70 +340,47 @@ def rmatvec(b):
 
     # Solve with preconditioner
     matvec_count = [0]
-    x, info = solver(A, b, M=precond, x0=x0, tol=tol)
+    if solver in [cg, cgs, bicg, bicgstab, gmres, qmr]:
+        x, info = solver(A, b, M=precond, x0=x0, rtol=tol)
+    else:
+        x, info = solver(A, b, M=precond, x0=x0, tol=tol)
 
     assert_equal(info, 0)
     assert_normclose(case.A.dot(x), b, tol)
 
     # Solution should be nearly instant
-    assert_(matvec_count[0] <= 3, repr(matvec_count))
+    assert matvec_count[0] <= 3
 
 
 @pytest.mark.parametrize("case", [params.Poisson1D, params.Poisson2D])
-def test_precond_inverse(case):
-    for solver in params.solvers:
-        if solver in case.skip:
-            continue
-        if solver is qmr:
-            continue
+@pytest.mark.parametrize('solver', params.solvers)
+def test_precond_inverse(case, solver):
+    if (solver in case.skip) or (solver is qmr):
+        pass
+    else:
         with suppress_warnings() as sup:
             sup.filter(DeprecationWarning, ".*called without specifying.*")
             check_precond_inverse(solver, case)
 
 
-def test_reentrancy():
-    non_reentrant = [cg, cgs, bicg, bicgstab, gmres, qmr]
-    reentrant = [lgmres, minres, gcrotmk, tfqmr]
-    for solver in reentrant + non_reentrant:
-        with suppress_warnings() as sup:
-            sup.filter(DeprecationWarning, ".*called without specifying.*")
-            _check_reentrancy(solver, solver in reentrant)
-
-
-def _check_reentrancy(solver, is_reentrant):
-    def matvec(x):
-        A = np.array([[1.0, 0, 0], [0, 2.0, 0], [0, 0, 3.0]])
-        y, info = solver(A, x)
-        assert_equal(info, 0)
-        return y
-    b = np.array([1, 1./2, 1./3])
-    op = LinearOperator((3, 3), matvec=matvec, rmatvec=matvec,
-                        dtype=b.dtype)
-
-    if not is_reentrant:
-        assert_raises(RuntimeError, solver, op, b)
-    else:
-        y, info = solver(op, b)
-        assert_equal(info, 0)
-        assert_allclose(y, [1, 1, 1])
-
-
-@pytest.mark.parametrize("solver", [cg, cgs, bicg, bicgstab, gmres, qmr, lgmres, gcrotmk])
+@pytest.mark.parametrize("solver", [cg, cgs, bicg, bicgstab, gmres, qmr,
+                                    lgmres, gcrotmk])
```

Comment:
This list is duplicated at least twice, and we should just turn the whole thing into a parametrized fixture, IMO.

(Suggestion dependent on others further up)
```suggestion
```